src/interface/variable.cpp

```diff
@@ -85,28 +86,42 @@ CellVariable<T>::AllocateCopy(const bool allocComms, std::weak_ptr<MeshBlock> wp
 /// Initialize a 6D variable
 template <typename T>
 void CellVariable<T>::allocateComms(std::weak_ptr<MeshBlock> wpmb) {
-  // set up fluxes
   std::string base_name = label();
+
+  // TODO(JMM): Note that this approach assumes LayoutRight. Otherwise
+  // the stride will mess up the types.
+
+  // Compute size of unified flux_data object and create it. A unified
+  // flux_data_ object reduces the number of memory allocations per
+  // variable per meshblock from 5 to 3.
+  int n_outer = 0;
   if (IsSet(Metadata::Independent)) {
-    flux[X1DIR] = ParArrayND<T>(base_name + ".fluxX1", GetDim(6), GetDim(5), GetDim(4),
-                                GetDim(3), GetDim(2), GetDim(1));
-    if (GetDim(2) > 1)
-      flux[X2DIR] = ParArrayND<T>(base_name + ".fluxX2", GetDim(6), GetDim(5), GetDim(4),
-                                  GetDim(3), GetDim(2), GetDim(1));
-    if (GetDim(3) > 1)
-      flux[X3DIR] = ParArrayND<T>(base_name + ".fluxX3", GetDim(6), GetDim(5), GetDim(4),
-                                  GetDim(3), GetDim(2), GetDim(1));
+    n_outer += (GetDim(1) > 1) + (GetDim(2) > 1) + (GetDim(3) > 1);
```

Comment:
```suggestion
  flux_data_ = ParArray7D<T>(base_name + ".flux_data", wpmb->pmy_mesh->ndim, GetDim(6),
                               GetDim(5), GetDim(4), GetDim(3), GetDim(2), GetDim(1));
```

Note, I haven't tried this change yet and I noticed the new `pmb->pmy_mesh != nullptr` below.
Can we call this function without a Mesh? then using `ndim` may not work after all.

Reply:
I wanted to keep this decoupled from `mesh` as much as possible. Variables are created without an underlying mesh in unit testing, and I think in general, it's a good idea to decouple the infrastructure as much as we can to enable self-contained tests. Relying on the existence of a mesh pointer will break the current unit tests.

Reply:
that said, I can see doing:
```C++
int n_outer = 1;
n_outer += (GetDim(2) > 1) + (GetDim(3) > 1);
```
to make it more clear that `n_outer` is always >= 1.

Alternatively I could add an assert to check that `n_outer >= 1` before creating the flux array.

Reply:
Resolved as discussed above